### DISCUSSION TITLE
Lesson 13: Integration test

### DISCUSSION THREAD

**Discussion by haraslub** [smartcontractkit/full-blockchain-solidity-course-py#514] [Q&A] [answered]
Hello,
I would like to ask you for explanation, please. I have gone through integration test and @PatrickAlphaC using the following approach in order to get balance of the token:
```Python
starting_balance = dapp_token.balanceOf(account.address)
```
The same approach is used then in the assertion part:
```Python
assert (
        dapp_token.balanceOf(account.address)
        == amount_token_to_issue + starting_balance
    )
```
**My question is, how is it possible to get non-zero values of balance even though there is a transfer function used once stakeTokens() function is applied (see the code below)? I would expect zero balance as the token was transferred to the token farm.**
```Solidity
 function stakeTokens(uint256 _amount, address token) public {
        // Require amount greater than 0
        require(_amount > 0, "amount cannot be 0");
        require(tokenIsAllowed(token), "Token currently isn't allowed");
        updateUniqueTokensStaked(msg.sender, token);
        IERC20(token).transferFrom(msg.sender, address(this), _amount);
        stakingBalance[token][msg.sender] =
            stakingBalance[token][msg.sender] +
            _amount;
        if (uniqueTokensStaked[msg.sender] == 1) {
            stakers.push(msg.sender);
        }
    }
```
Thank you very much in advance.

**Comment by cromewar** [comment, marked as answer]
Hello @haraslub that's quite a good question and I'm not quite sure I will answer correctly so @PatrickAlphaC could  you give us a hand on this one?

As far I understand I would pretty much say that it happens due to how we deploy the token, as we give the balance to the msg-sender at the constructor, so if you stake the tokens using the same account, you don't actually stake all of them so the account will never be empty. But again I'm not sure, let's hope @PatrickAlphaC could give us more light on this.


![image](https://user-images.githubusercontent.com/20199180/145132406-96f72554-d61c-4e52-b56d-921600619804.png)

**Reply by haraslub** [reply]
Hello @cromewar thank you very much for your feedback. I hope we will have clarity on that. :-)

**Reply by PatrickAlphaC** [reply]
@cromewar is 100% correct!

When we mint the token, _someone_ has to be the original owner of all the tokens. And in our contract, we have the address that deployed the contract own all the tokens.

**Reply by haraslub** [reply]
@PatrickAlphaC thank you very much for your confirmation/explanation.
